### PR TITLE
Stop confusing inner and outer outputs in connection_pattern()

### PR DESCRIPTION
**Pull request by carriepl**
Based on the code and comments, the implementation of scan.connection_pattern appears to confuse inner and outer outputs. The method `get_input_pos()`, means to receive the index of an *inner* output is called with the indices of *outer* outputs. This means that if the scan has mitmot outputs with many output taps, some of its recurrent outputs might never be processed.

I tried to come up with a test case for this but couldn't find one where there would be a measurable, observable difference between the connection_pattern before this PR and after. I have a test case which I haven't included where the code behavior is different but the final connection_pattern is identical.

### CODE REVIEW

**Comment by abergeron** [comment]
This seems cleaner, but I honestly have no idea if it's better or not.

File: theano/scan_module/scan_op.py [hunk 1]
```diff
@@ -1471,35 +1471,54 @@ def _get_inner_inps_idx(outer_iidx):
...
+            result[output_offset] = input_offset
+            input_offset += 1
+            output_offset += 1
+
+        # Process shared inputs/outputs
+        input_offset += self.n_nit_sot
+        for i in range(self.n_shared_outs):
+            result[output_offset + i] = input_offset
+            input_offset += 1
+            output_offset += 1
```
Comment (lamblin) [inline theano/scan_module/scan_op.py]: I don't understand that part. It looks like you are incrementing `output_offset` but also adding it `i`, meaning that only every second value in `result` gets set.
Maybe that is on purpose, but it is confusing, and not documented.

Comment (carriepl) [inline theano/scan_module/scan_op.py]: @lamblin : That is not on purpose. You are absolutely right.

**Comment by carriepl** [comment]
Fixed the issue and added some tests to make sure.

**Comment by lamblin** [comment]
There is a failing test in FAST_COMPILE

**Pull request merged**